File: src/main/java/dev/rodolphe/accesscontrol/signaling/SignalingHub.java
package dev.rodolphe.accesscontrol.signaling;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.TaskScheduler;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * The in-memory relay between intercoms and residents: who is connected, which calls are in flight,
 * and who may say what to whom.
 *
 * <p>Nothing here is persisted. On a restart every client reconnects and the state rebuilds itself —
 * a call that was ringing simply stops ringing, which is the right outcome anyway.
 *
 * <p>A singleton by necessity, not by convenience. Two instances would mean an intercom registered in
 * one and a resident in the other, and nobody ever reaching anybody. Spring's default scope gives
 * that guarantee for free, where the Kotlin server obtained it by creating the object once in
 * {@code module()} and passing the same reference everywhere.
 *
 * <p>Being a shared singleton crossed by every connection at once, it holds only concurrent maps, and
 * every state transition is decided by a compare-and-set rather than by a read followed by a write.
 */
@Component
public class SignalingHub {

    private final Map<String, ClientConnection> residents = new ConcurrentHashMap<>(); // userId -> conn
    private final Map<String, ClientConnection> intercoms = new ConcurrentHashMap<>(); // buildingId -> conn
    private final Map<String, CallState> calls = new ConcurrentHashMap<>();            // callId -> state

    private final TaskScheduler scheduler;
    private final long ringTimeoutMs;

    /**
     * The scheduler replaces the {@code CoroutineScope} the Kotlin hub was handed. Its only job is the
     * "nobody picked up" timer; injecting it rather than creating one keeps the timing controllable
     * from a test, which is how the expiry path gets covered at all.
     */
    public SignalingHub(TaskScheduler scheduler,
                        @Value("${syekso.signaling.ring-timeout-ms:30000}") long ringTimeoutMs) {
        this.scheduler = scheduler;
        this.ringTimeoutMs = ringTimeoutMs;
    }

    public void registerResident(String userId, ClientConnection connection) {
        residents.put(userId, connection);
    }

    public void registerIntercom(String buildingId, ClientConnection connection) {
        intercoms.put(buildingId, connection);
    }

    /** Drops a closed connection and abandons any call that depended on it. */
    public void unregister(ClientConnection connection) {
        residents.values().removeIf(known -> known == connection);
        intercoms.values().removeIf(known -> known == connection);
        calls.entrySet().removeIf(entry -> {
            CallState state = entry.getValue();
            boolean broken = !residents.containsKey(state.residentUserId())
                    || !intercoms.containsKey(state.buildingId());
            if (broken) {
                state.cancelTimeout();
            }
            return broken;
        });
    }

    /** An intercom rings a resident. One call at a time per intercom. */
    public void onRingCall(String buildingId, SignalingMessage.Ring message) {
        String targetUserId = message.targetUserId();
        if (targetUserId == null) {
            return;
        }
        if (calls.values().stream().anyMatch(state -> state.buildingId().equals(buildingId))) {
            sendToIntercom(buildingId, new SignalingMessage.ErrorMsg(message.callId(), "Appel déjà en cours"));
            return;
        }
        ClientConnection resident = residents.get(targetUserId);
        if (resident == null) {
            sendToIntercom(buildingId, new SignalingMessage.ErrorMsg(message.callId(), "Résident indisponible"));
            return;
        }

        CallState state = new CallState(buildingId, targetUserId);
        calls.put(message.callId(), state);
        state.setTimeout(scheduler.schedule(
                () -> expire(message.callId()), Instant.now().plusMillis(ringTimeoutMs)));

        resident.send(new SignalingMessage.Ring(message.callId(), targetUserId, message.doorName()));
    }

    /** Nobody answered in time. Does nothing if the call was accepted in the meantime. */
    private void expire(String callId) {
        CallState state = calls.get(callId);
        if (state == null || !state.markExpiredIfRinging()) {
            return;
        }
        calls.remove(callId);
        sendToIntercom(state.buildingId(), new SignalingMessage.ErrorMsg(callId, "Pas de réponse"));
        sendToResident(state.residentUserId(), new SignalingMessage.ErrorMsg(callId, "TIMED_OUT"));
    }

    /** The resident answered: ringing becomes in-call, and the intercom is told to make its offer. */
    public void onAcceptCall(String residentUserId, SignalingMessage.Accept message) {
        CallState state = calls.get(message.callId());
        if (state == null
                || !state.residentUserId().equals(residentUserId)
                || !state.markInCallIfRinging()) {
            sendToResident(residentUserId, new SignalingMessage.ErrorMsg(message.callId(), "Appel expiré"));
            return;
        }
        state.cancelTimeout();
        sendToIntercom(state.buildingId(), new SignalingMessage.Accept(message.callId()));
    }

    /** Open the door mid-call. Deliberately does not end the call — the conversation continues. */
    public void onOpenCall(String residentUserId, SignalingMessage.Open message) {
        CallState state = calls.get(message.callId());
        if (state == null || !state.residentUserId().equals(residentUserId)) {
            sendToResident(residentUserId, new SignalingMessage.ErrorMsg(message.callId(), "Appel expiré"));
            return;
        }
        ClientConnection intercom = intercoms.get(state.buildingId());
        if (intercom == null) {
            sendToResident(residentUserId, new SignalingMessage.ErrorMsg(message.callId(), "Interphone hors ligne"));
            return;
        }
        intercom.send(new SignalingMessage.Open(message.callId()));
    }

    public void onDeclineCall(String residentUserId, SignalingMessage.Decline message) {
        CallState state = calls.remove(message.callId());
        if (state == null) {
            return;
        }
        state.cancelTimeout();
        sendToIntercom(state.buildingId(), new SignalingMessage.Decline(message.callId()));
    }

    /** The real BLE outcome, relayed back to the resident. The call stays up. */
    public void onOpenResultReported(String buildingId, SignalingMessage.OpenResult message) {
        CallState state = calls.get(message.callId());
        if (state == null || !state.buildingId().equals(buildingId)) {
            return;
        }
        sendToResident(state.residentUserId(), message);
    }

    /** Hangup from either side: end the call and tell the other peer. */
    public void onHangupCall(String callId, boolean fromResident) {
        CallState state = calls.remove(callId);
        if (state == null) {
            return;
        }
        state.cancelTimeout();
        if (fromResident) {
            sendToIntercom(state.buildingId(), new SignalingMessage.Hangup(callId));
        } else {
            sendToResident(state.residentUserId(), new SignalingMessage.Hangup(callId));
        }
    }

    /** Media negotiation from the resident, passed through without being read. */
    public void relayFromResident(String residentUserId, SignalingMessage.WebRtc message) {
        CallState state = calls.get(message.callId());
        if (state == null || !state.residentUserId().equals(residentUserId)) {
            return;
        }
        sendToIntercom(state.buildingId(), message);
    }

    /** Media negotiation from the intercom, passed through without being read. */
    public void relayFromIntercom(String buildingId, SignalingMessage.WebRtc message) {
        CallState state = calls.get(message.callId());
        if (state == null || !state.buildingId().equals(buildingId)) {
            return;
        }
        sendToResident(state.residentUserId(), message);
    }

    private void sendToResident(String userId, SignalingMessage message) {
        ClientConnection connection = residents.get(userId);
        if (connection != null) {
            connection.send(message);
        }
    }

    private void sendToIntercom(String buildingId, SignalingMessage message) {
        ClientConnection connection = intercoms.get(buildingId);
        if (connection != null) {
            connection.send(message);
        }
    }
}

File: src/main/kotlin/dev/rodolphe/accesscontrol/signaling/SignalingHub.kt
package dev.rodolphe.accesscontrol.signaling

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.ConcurrentHashMap

/** One connected client (resident or intercom). [rawSend] serializes+sends one message; the mutex
 *  serializes concurrent sends to the same socket so frames don't interleave. */
class ClientConnection(
    val id: String,
    private val rawSend: suspend (SignalingMessage) -> Unit,
) {
    private val sendMutex = Mutex()
    suspend fun send(msg: SignalingMessage) = sendMutex.withLock { rawSend(msg) }
}

enum class CallStatus { RINGING, IN_CALL }

class CallState(
    val buildingId: String,
    val residentUserId: String,
    var status: CallStatus,
    var timeoutJob: Job? = null,
)

/** In-memory relay. Not persisted: on backend restart, clients reconnect. */
class SignalingHub(
    private val scope: CoroutineScope,
    private val ringTimeoutMs: Long = 30_000,
) {
    private val residents = ConcurrentHashMap<String, ClientConnection>()   // userId -> conn
    private val intercoms = ConcurrentHashMap<String, ClientConnection>()   // buildingId -> conn
    private val calls = ConcurrentHashMap<String, CallState>()              // callId -> state

    fun registerResident(userId: String, conn: ClientConnection) { residents[userId] = conn }
    fun registerIntercom(buildingId: String, conn: ClientConnection) { intercoms[buildingId] = conn }

    /** Remove a dropped connection and cancel any call that depended on it. */
    fun unregister(conn: ClientConnection) {
        residents.entries.removeIf { it.value === conn }
        intercoms.entries.removeIf { it.value === conn }
        calls.entries.removeIf { (_, state) ->
            val broken = residents[state.residentUserId] == null || intercoms[state.buildingId] == null
            if (broken) state.timeoutJob?.cancel()
            broken
        }
    }

    suspend fun onRingCall(buildingId: String, msg: SignalingMessage.Ring) {
        val targetUserId = msg.targetUserId ?: return
        // One call at a time per intercom.
        if (calls.values.any { it.buildingId == buildingId }) {
            intercoms[buildingId]?.send(SignalingMessage.ErrorMsg(msg.callId, "Appel déjà en cours"))
            return
        }
        val resident = residents[targetUserId]
        if (resident == null) {
            intercoms[buildingId]?.send(SignalingMessage.ErrorMsg(msg.callId, "Résident indisponible"))
            return
        }
        val state = CallState(buildingId, targetUserId, CallStatus.RINGING)
        calls[msg.callId] = state
        state.timeoutJob = scope.launch {
            delay(ringTimeoutMs)
            if (calls[msg.callId]?.status == CallStatus.RINGING) {
                calls.remove(msg.callId)
                intercoms[buildingId]?.send(SignalingMessage.ErrorMsg(msg.callId, "Pas de réponse"))
                residents[targetUserId]?.send(SignalingMessage.ErrorMsg(msg.callId, "TIMED_OUT"))
            }
        }
        resident.send(SignalingMessage.Ring(msg.callId, targetUserId, msg.doorName))
    }

    /** Resident answered: RINGING -> IN_CALL, relay Accept to the intercom (which then creates the offer). */
    suspend fun onAcceptCall(residentUserId: String, msg: SignalingMessage.Accept) {
        val state = calls[msg.callId]
        if (state == null || state.residentUserId != residentUserId || state.status != CallStatus.RINGING) {
            residents[residentUserId]?.send(SignalingMessage.ErrorMsg(msg.callId, "Appel expiré"))
            return
        }
        state.status = CallStatus.IN_CALL
        state.timeoutJob?.cancel()
        intercoms[state.buildingId]?.send(SignalingMessage.Accept(msg.callId))
    }

    /** Open the door during a call: relay OPEN to the intercom. Does NOT end the call. */
    suspend fun onOpenCall(residentUserId: String, msg: SignalingMessage.Open) {
        val state = calls[msg.callId]
        if (state == null || state.residentUserId != residentUserId) {
            residents[residentUserId]?.send(SignalingMessage.ErrorMsg(msg.callId, "Appel expiré"))
            return
        }
        val intercom = intercoms[state.buildingId]
        if (intercom == null) {
            residents[residentUserId]?.send(SignalingMessage.ErrorMsg(msg.callId, "Interphone hors ligne"))
            return
        }
        intercom.send(SignalingMessage.Open(msg.callId))
    }

    suspend fun onDeclineCall(residentUserId: String, msg: SignalingMessage.Decline) {
        val state = calls.remove(msg.callId) ?: return
        state.timeoutJob?.cancel()
        intercoms[state.buildingId]?.send(SignalingMessage.Decline(msg.callId))
    }

    /** Real BLE result relayed back to the resident. Call stays alive (talk continues). */
    suspend fun onOpenResultReported(buildingId: String, msg: SignalingMessage.OpenResult) {
        val state = calls[msg.callId] ?: return
        if (state.buildingId != buildingId) return
        residents[state.residentUserId]?.send(msg)
    }

    /** Hangup from either side: end the call, relay to the other peer. */
    suspend fun onHangupCall(callId: String, fromResident: Boolean) {
        val state = calls.remove(callId) ?: return
        state.timeoutJob?.cancel()
        val target = if (fromResident) intercoms[state.buildingId] else residents[state.residentUserId]
        target?.send(SignalingMessage.Hangup(callId))
    }

    /** Pure media pass-through from the resident to the intercom (Answer, IceCandidate). */
    suspend fun relayFromResident(residentUserId: String, msg: SignalingMessage.WebRTC) {
        val state = calls[msg.callId] ?: return
        if (state.residentUserId != residentUserId) return
        intercoms[state.buildingId]?.send(msg)
    }

    /** Pure media pass-through from the intercom to the resident (Offer, IceCandidate). */
    suspend fun relayFromIntercom(buildingId: String, msg: SignalingMessage.WebRTC) {
        val state = calls[msg.callId] ?: return
        if (state.buildingId != buildingId) return
        residents[state.residentUserId]?.send(msg)
    }
}
